apache/samza
# Add new configuration allowing to keep processing when there are fatal exceptions or timeout

## Changes

Introduce configurations allowing the job to keep processing when there is exception or timeout in the commit stage

- Add new configurations 
  1. `task.commit.skip.commit.during.failures.enabled` with default to `false`. When the config is set to true, decouple the fatal checkpoint errors and commit timeouts from data processing
  2. `task.commit.skip.commit.exception.max.limit` with default to `5`. Note that the commit period config `task.commit.ms` defaults to `60s`, the worst scenario would be the job is shut down after 5 exceptions (5 mins). 
  3. `task.commit.skip.commit.timeout.max.limit` with default to `2`. Note that the commit timeout config `task.commit.timeout.ms` defaults to `30mins`, the worst scenario would be the job is shut down after 2 timeouts (60 mins). 
- Fix / update the metrics for commit
  - Update `commits-timed-out` and `commits-skipped` to Counter  
  - Create new metric `commit-exceptions` denoting the counter of the exceptions during commit
  - Create new metric `last-async-commit-ns` denoting the finish timestamp last successfully async 
- Update error / timeout handling mechanism with the config setting to true. 

---

## Tests
- [x] executed `./gradlew build` 
- [x] unit test in `TestTaskInstance.scala`
- [x] E2E tests (Details below) 

Preparation steps:
1. Introduce new config `task.commit.test.commit.exception.enabled` for testing only
  - When the config is set to true, a `RuntimeException` will be thrown during the commit 
2. Introduce new config `task.commit.test.commit. timeout.enabled` for testing only
 - When the config is set to true, the async stage of the commit will contains a TimeoutFuture that put the thread into sleep

E2E testing: 
1. Build a stateful job without setting any configs introduced in this PR  -> **job works as normal**
2. Set following configs -> **job works as normal**
  - `task.commit.skip.commit.during.failures.enabled=true` 
3. Set the following configs 
  - `task.commit.skip.commit.during.failures.enabled=true`
  - `task.commit.skip.commit.exception.max.limit=2`
  - `task.commit.test.commit.exception.enabled=true`
And the job failed with the exception below (**expected**) 
```
“Caused by: org.apache.samza.SamzaException: Unrecoverable error during pending commit for taskName: Partition 8. Exception Counter: 3”
“Caused by: java.util.concurrent.CompletionException: java.lang.RuntimeException: Exception between trimFuture and whenCompleteAsync”
```
4. Set the following configs 
  - `task.commit.skip.commit.during.failures.enabled=true`
  - `task.commit.skip.commit.timeout.max.limit=2`
  - `task.commit.test.commit.timeout.enabled=true`
  - `job.container.thread.pool.size=16` (If this is not set, it's default to doing commit in single thread) 
The job failed with the exception below (**expected**) 
```
org.apache.samza.SamzaException: org.apache.samza.SamzaException: Timeout waiting for pending commit for taskName: Partition 0 to finish. 29998 ms have elapsed since the pending commit started. Max allowed commit delay is 8000 ms and commit timeout beyond that is 15000 ms. Timeout Counter: 3
```
